data-prepper-plugins/anomaly-detector-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/anomalydetector/AnomalyDetectorProcessor.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.anomalydetector;
+
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.configuration.PluginModel;
+import org.opensearch.dataprepper.model.configuration.PluginSetting;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.processor.AbstractProcessor;
+import org.opensearch.dataprepper.model.processor.Processor;
+import org.opensearch.dataprepper.model.plugin.PluginFactory;
+
+import java.util.Collection;
+import java.util.List;
+
+@DataPrepperPlugin(name = "anomaly_detector", pluginType = Processor.class, pluginConfigurationType = AnomalyDetectorProcessorConfig.class)
+public class AnomalyDetectorProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    public static final String DEVIATION_KEY = "deviation_from_expected";
+    public static final String GRADE_KEY = "grade";
+
+    private List<String> keys;
+    private AnomalyDetectorMode mode;
+    AnomalyDetectorProcessorConfig anomalyDetectorProcessorConfig;
+
+    @DataPrepperPluginConstructor
+    public AnomalyDetectorProcessor(final AnomalyDetectorProcessorConfig anomalyDetectorProcessorConfig, final PluginMetrics pluginMetrics, final PluginFactory pluginFactory) {
+        super(pluginMetrics);
+        this.anomalyDetectorProcessorConfig = anomalyDetectorProcessorConfig;
+        keys = anomalyDetectorProcessorConfig.getKeys();
+        mode = loadAnomalyDetectorMode(pluginFactory);
+        mode.initialize(keys);
+    }
+
+    private AnomalyDetectorMode loadAnomalyDetectorMode(final PluginFactory pluginFactory) {
+        final PluginModel modeConfiguration = anomalyDetectorProcessorConfig.getDetectorMode();
+        final PluginSetting modePluginSetting = new PluginSetting(modeConfiguration.getPluginName(), modeConfiguration.getPluginSettings());
+        return pluginFactory.loadPlugin(AnomalyDetectorMode.class, modePluginSetting);
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        return mode.handleEvents(records);
+    }
+
+
+    @Override
+    public void prepareForShutdown() {
+
+    }
```

Comment:
If data prepper is shutdown to update the configuration or some other operation, do we need to preserve the Random Cut Forest? Will the underlying library handle this? Do we need to save it to disk or do we expect restarting DP to clear the state?

Reply:
Let me take the question. "do we need to preserve the Random Cut Forest"? Not necessarily, but you can choose to save it. If you do not preserve the forest, then the forest will relearn on the new data (which I believe is the intention here, and appropriate as a first option for high throughput data). If you chose to save the forest, the library has ser/De routines. I  believe the intention here is a generic "prepareForShutdown" which can be extended to other processors/different mode of the same processor.

Reply:
For now (Initial implementation), we are not going to support saving the tree across pipeline restarts.